server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -141,77 +145,112 @@ public void clusterChanged(final ClusterChangedEvent event) {
-                enforceRecoverAfterTime = true;
-                reason = "recover_after_time was set to [" + recoverAfterTime + "]";
-            } else if (expectedDataNodes <= nodes.getDataNodes().size()) {
-                // expected is set and satisfied so recover immediately
-                enforceRecoverAfterTime = false;
-                reason = "";
-            } else {
-                // expected is set but not satisfied so wait until it is satisfied or times out
-                enforceRecoverAfterTime = true;
-                reason = "expecting [" + expectedDataNodes + "] data nodes, but only have [" + nodes.getDataNodes().size() + "]";
-            }
-            performStateRecovery(enforceRecoverAfterTime, reason);
+            return;
+        }
+
+        // At this point, we know the state is not recovered and this node is qualified for state recovery
+        // But we still need to check whether a previous one is running already
+        final long currentTerm = state.term();
+        if (pendingStateRecovery.term < currentTerm) {
+            // Always start a new state recovery if the master term changes
+            // If there is a previous one still waiting, both will run but at most one of them will
+            // actually make changes to cluster state
+            pendingStateRecovery = new PendingStateRecovery(currentTerm);
         }
+        assert pendingStateRecovery.term == currentTerm;
+        pendingStateRecovery.maybeStart(nodes.getDataNodes().size());
     }
 
-    private void performStateRecovery(final boolean enforceRecoverAfterTime, final String reason) {
-        if (enforceRecoverAfterTime && recoverAfterTime != null) {
-            if (scheduledRecovery.compareAndSet(false, true)) {
-                logger.info("delaying initial state recovery for [{}]. {}", recoverAfterTime, reason);
-                threadPool.schedule(new AbstractRunnable() {
-                    @Override
-                    public void onFailure(Exception e) {
-                        logger.warn("delayed state recovery failed", e);
-                        resetRecoveredFlags();
-                    }
+    class PendingStateRecovery {
+        private final long term;
+        @Nullable
+        private SubscribableListener<Void> recoveryPlanned;
 
-                    @Override
-                    protected void doRun() {
-                        if (recoveryInProgress.compareAndSet(false, true)) {
-                            logger.info("recover_after_time [{}] elapsed. performing state recovery...", recoverAfterTime);
-                            runRecovery();
-                        }
-                    }
-                }, recoverAfterTime, threadPool.generic());
+        PendingStateRecovery(long term) {
+            this.term = term;
+        }
+
+        void maybeStart(int dataNodeSize) {
+            final SubscribableListener<Void> thisRecoveryPlanned;
```

Comment:
On reflection doing this with a `SubscribableListener` actually seems more awkward than just sticking with the original `threadPool.schedule`.

Reply:
Sure I changed it to use just `threadPool.schedule`. It might make  `SubscribableListener` more suitable for this use case if it does not have to use Exception to signal timeout.